codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,506 @@
+ *
+ * <p>The HTTP/2 standard allows for an endpoint to limit the maximum number of concurrently active streams via the
+ * {@code SETTINGS_MAX_CONCURRENT_STREAMS} setting. When this limit is reached, no new streams can be created. However,
+ * the {@link Http2FrameCodec} can be build with
+ * {@link Http2FrameCodecBuilder#encoderEnforceMaxConcurrentStreams(boolean)} enabled, in which case a new stream and
+ * its associated frames will be buffered until either the limit is increased or an active stream is closed. It's,
+ * however, possible that a buffered stream will never become active. That is, the channel might
+ * get closed or a GO_AWAY frame might be received. In the first case, all writes of buffered streams will fail with a
+ * {@link Http2ChannelClosedException}. In the second case, all writes of buffered streams with an identifier less than
+ * the last stream identifier of the GO_AWAY frame will fail with a {@link Http2GoAwayException}.
+ *
+ * <h3>Error Handling</h3>
+ *
+ * Exceptions and errors are propagated via {@link ChannelInboundHandler#exceptionCaught}. Exceptions that apply to
+ * a specific HTTP/2 stream are wrapped in a {@link Http2FrameStreamException} and have the corresponding
+ * {@link Http2FrameStream} object attached.
+ *
+ * <h3>Reference Counting</h3>
+ *
+ * Some {@link Http2StreamFrame}s implement the {@link ReferenceCounted} interface, as they carry
+ * reference counted objects (e.g. {@link ByteBuf}s). The frame codec will call {@link ReferenceCounted#retain()} before
+ * propagating a reference counted object through the pipeline, and thus an application handler needs to release such
+ * an object after having consumed it. For more information on reference counting take a look at
+ * http://netty.io/wiki/reference-counted-objects.html
+ *
+ * <h3>HTTP Upgrade</h3>
+ *
+ * Server-side HTTP to HTTP/2 upgrade is supported in conjunction with {@link Http2ServerUpgradeCodec}; the necessary
+ * HTTP-to-HTTP/2 conversion is performed automatically.
  */
 @UnstableApi
-public class Http2FrameCodec extends ChannelDuplexHandler {
+public class Http2FrameCodec extends Http2ConnectionHandler {
+
+    private static final InternalLogger LOG = InternalLoggerFactory.getInstance(Http2FrameCodec.class);
 
-    private static final Http2FrameLogger HTTP2_FRAME_LOGGER = new Http2FrameLogger(INFO, Http2FrameCodec.class);
+    private final PropertyKey streamKey;
 
-    private final Http2ConnectionHandler http2Handler;
-    private final boolean server;
+    private final Integer initialFlowControlWindowSize;
 
     private ChannelHandlerContext ctx;
-    private ChannelHandlerContext http2HandlerCtx;
 
-    /**
-     * Construct a new handler.
-     *
-     * @param server {@code true} this is a server
-     */
+    /** Number of buffered streams if the {@link StreamBufferingEncoder} is used. **/
+    private int numBufferedStreams;
+
     public Http2FrameCodec(boolean server) {
```

Comment:
should we encourage a consistent initialization approach via the builder and not expose these constructors publicly?